internal/retry/request.go

```diff
@@ -0,0 +1,248 @@
+	})
+}
+
+func MakeOptions(os ...Option) *Options {
+	opts := &Options{
+		maxRetries:   DefaultMaxRetries,
+		initialDelay: DefaultInitialDuration,
+		sleep:        time.Sleep,
+		backoff:      DefaultBackoff,
+	}
+	for _, o := range os {
+		o.Apply(opts)
+	}
+	return opts
+}
+
+type Request struct {
+	client   func(*http.Request) (*http.Response, error)
+	r        *http.Request
+	o        *Options
+	attempts int
+	done     bool
+	delay    time.Duration
+}
+
+func NewRequest(r *http.Request, client func(*http.Request) (*http.Response, error), o *Options) *Request {
+	return &Request{
+		client: client,
+		r:      r,
+		o:      o,
+	}
+}
+
+func (r *Request) Done() bool {
+	return r.done || r.attempts >= r.o.maxRetries
+}
+
+func (r *Request) onError(err error) (*http.Response, error) {
+	r.done = true
+	return nil, err
+}
+
+func (r *Request) onDone(resp *http.Response, err error) (*http.Response, error) {
+	r.done = true
+	return resp, err
+}
+
+func (r *Request) Do() (*http.Response, error) {
+	if r.Done() {
+		// Already done, so don't attempt any more.
+		return nil, ErrorNoMoreAttempts
+	}
+	if r.attempts > 0 {
+		// This is a retry!
+		if r.delay > 0 {
+			// Wait if we have a delay
+			r.o.sleep(r.delay)
+		}
+		// Update the delay
+		r.delay = r.o.backoff(r.delay)
+	}
+	// Bump the number of attempts
+	r.attempts++
+
+	// Issue the request
+	resp, err := r.client(r.r)
+	if err != nil {
+		// Return the response and the error if the client returned an error
+		// TODO: pass err to the strategy funcs.
+		return r.onDone(resp, err)
+	}
+
+	// Immediate success if the request is 2xx or 3xx
+	if http.StatusOK <= resp.StatusCode && resp.StatusCode < http.StatusBadRequest {
+		return r.onDone(resp, err)
+	}
+
+	if r.o.retryAfter != nil {
+		// Check if the Retry-After header is set
+		d := r.o.retryAfter(resp)
+		if d != 0 {
+			// We have the Retry-After header so set the delay and return
+			r.delay = d
+			return resp, err
+		}
+	}
+
+	// Check each of the retryStrategyFuncs to see if we should retry
+	s := NoRetry
+	var errS error
+	for _, fn := range r.o.retryStrategyFuncs {
+		s, errS = fn(resp)
+		if errS != nil {
+			return r.onError(errS)
+		}
+		if s != NoRetry {
+			// We found a strategy that required a retry
+			// TODO: we may want to find the retry needing the longest delay
+			break
+		}
+	}
+	if s == NoRetry {
+		return r.onDone(resp, err)
+	}
+	// Only apply the strategy on the first iteration
+	if r.attempts == 1 && s == RetryWithInitialDelay {
+		r.delay = r.o.initialDelay
+	}
+
+	return resp, err
+}
+
+func NewRoundTripper(inner http.RoundTripper, o ...Option) http.RoundTripper {
+	return &roundTripper{
+		inner: inner,
+		o:     MakeOptions(o...),
+	}
+}
+
+type roundTripper struct {
+	inner http.RoundTripper
+	o     *Options
+}
+
+func (rt *roundTripper) RoundTrip(r *http.Request) (*http.Response, error) {
+	var resp *http.Response
+	var err error
+	rr := NewRequest(r, rt.inner.RoundTrip, rt.o)
+	for !rr.Done() {
```

Comment:
in what case would this be done already?

Reply:
Never, immediately after instantiation.

The forloop semantics are convenient for being able to make the retry library testable.

Technically you could implement this as:

```go
for {
  resp, err := rr.Do()
  if rr.Done() {
    break
  }
}
```

I'll add a comment making this clearer.

Reply:
ack. I misread the code here actually. thanks!